src/common/utils/oidc/helper.go

```diff
@@ -305,14 +306,28 @@ func userInfoFromIDToken(ctx context.Context, token *Token, setting models.OIDCS
 	if err != nil {
 		return nil, err
 	}
-	return userInfoFromClaims(idt, setting.GroupsClaim)
+
+	return userInfoFromClaims(idt, setting.GroupsClaim, setting.UserClaim)
 }
 
-func userInfoFromClaims(c claimsProvider, g string) (*UserInfo, error) {
+func userInfoFromClaims(c claimsProvider, g, u string) (*UserInfo, error) {
 	res := &UserInfo{}
 	if err := c.Claims(res); err != nil {
 		return nil, err
 	}
+	if u != "" {
+		allClaims := make(map[string]interface{})
+		if err := c.Claims(&allClaims); err != nil {
+			return nil, err
+		}
+
+		if username, ok := allClaims[u].(string); !ok {
+			log.Warningf("OIDC. Failed to recover Username from claim. Claim '%s' is empty", u)
```

Comment:
It's also possible that claim is non-string.

Additionally, we'd better check the string is valid for a username, we may want to refactor and share the logic with this:
https://github.com/goharbor/harbor/blob/master/src/core/api/user.go#L598

Reply:
Hi @reasonerjt, in case the claim is not a string, it is not valid, the warning will be logged and the user won't be recovered (same case if the claim is not a string, does not exist, etc.). I might just change the warning message to tell that claim is empty or invalid.

About validating the user, I don't think this is the right moment to perform the user validation. No validation is done if the user is retrieved from the default claim, so I see no sense in doing the validation if it is retrieved from a customized claim. I thin the validation is performed later, when we try to onboard the user in https://github.com/goharbor/harbor/blob/426bce6ce4db66dde6e87ea3cf932798dea87174/src/core/controllers/oidc.go#L258, so the user is validated no matter if a custom user token is set, or the default token is used.

Reply:
I made a change to trigger an error instead of just logging a Warning in the log. So in case the user configures an invalid username claim (claim does not exist, or it is not a string), an error is raised now.

I like the idea of adding some validation, but it shouldn't be at this point, but on the userOnboard function, and outside the scope of this patch. in fact, even without this patch, it is possible to type a username like " " (a single blank) in the onboard screen, and it will be accepted.